### ISSUE TITLE
Model Evaluation Sub-package / package

### ISSUE THREAD

**Issue by nybyhansen** [dfds-data/dfds-ds-toolbox#22]
One idea for some functionalities to add in this package is some tooling around model evaluation, both for:

- individual model evaluation
- comparison of multiple models

Ideally we can use this tool to evaluate our models **also** in terms of model behaviour instead of only a flat metric (accuracy, mse etc.).
The tool should support initial model development (one model evaluation) and automatic retraining / deployment (comparison of multiple models)..

**Please, let's have a discussion of bricks and actual tools that could fit into this!**
When we have a set of desired functionalities we can start thinking of a good design and break it further down into tasks.. 😊

**Comment by KPLauritzen** [comment]
I like partial dependence plots to see how some changing features will impact predictions. 
https://scikit-learn.org/stable/modules/generated/sklearn.inspection.PartialDependenceDisplay.html#sklearn.inspection.PartialDependenceDisplay
This indicates something about what patterns the model has actually learned

**Comment by KPLauritzen** [comment]
Pinging @dfds-data/data-science-chapter for comments.

**Comment by KPLauritzen** [comment]
To be able to make any kind of general evaluator, we need an interface. We should either inherit something from sklearn or make our own AbstractBaseClass that borrows the same interface. @dingobar

**Comment by dingobar** [comment]
Is there a generic framework for this?

Suggest we create an abstract interface to our models, and then a standard method for comparing and evaluating models that implement that interface. In this way, we accomplish
- High level abstraction encourages maintainable code
- Framework for collaboration and joint decision making
- The interface can be reused for other features we decide to add to the package

```python
class ModelEvaluator:
    def compare(data: ..., models: List[ModelInterface]):
        ...
```

**Comment by kaqmak** [comment]
I may not be the most qualified for input on this, but some comments:
1. Are we aiming for something that is able to generalize to all types of models? E.g. A model evaluator that works for a time-series arima model, a Bayesian model, image classifier etc ?
2. Is the modelEvaluator supposed to be estimator agnostic? E:g. Is the aim of model comparison to compare two models that are based on the same type of estimator? If so, some more evaluator specific comparison can be carried out. E.g. for a Bayesian model, I would be interested in some sample metrics, effective sample size etc. Moreover, two types of estimators/models may require different scaling. E.g. one needs fitted to log(y) while one to y. Does a RMSE makes sense to use for comparing the two?

**Comment by nybyhansen** [comment]
I think we should definitely make use of interfaces here @dingobar and @KPLauritzen! I am not a fan of having this put on the model class as it violates SRP and might be more complex than this :blush: And I don't find functionality in scikit-learn that mimics this so I think we should only respect the model interface.

I have this idea in mind that I can try to elaborate on:
Having a design that allows us to set up a consolidated **validator** that consists of multiple **validationsteps** that allows us to set up a very custom and extendable validation scheme that can grow over time. Example of a validation step could be the simple flat metrics that we are used to, like _accuracy_.  Next one could be a validation step that requires all bins of the most important feature to not fall under the average performance with X standard deviation, and so on..

They could be chained together in a Pipeline like fashion but I would a validation step to also be "not required to pass" to allow the best model be the one passing the most validation steps. 

In terms of how to design this, I am not really sure. But the _chain of responsibility_ design patterns could serve as a good inspiration.

**Comment by nybyhansen** [comment]
@kaqmak 

1. We could but it might have different validators. Unstructured data like images, text and voice is fundamentally different from tabular so it makes sense to splitting them.
2. I hope that we could have it estimator agnostic as long as the estimator implemented inherits from the model interface, i.e. having the same design (fit and predict) and output (array, dataframe). We of course need to handle it in case that one model relies on standardized data or log transformed target.

**Comment by nybyhansen** [comment]
@KPLauritzen But that is already implemented in scikit-learn? Do you think of it as being a validation step, like "this feature needs to have this positive relationship with prediction" in order for the validation to pass? :blush:

**Comment by kaqmak** [comment]
@nybyhansen I think a case of comparing models where you have  the same estimator, and the same features could also relevant (but maybe not what you are trying to address here?). E.g.  You have had a trained model in production for a long time. Some new data comes in, and you decide to retrain. For some reason, the model is now different (e.g in terms of feature attribution). Flat metrics show more or less the same (or some are better than the old, some are worse) . In this case you would like to be notified of these, at first sight, invisible changes.

**Comment by nybyhansen** [comment]
@kaqmak yes indeed, that is my idea behind "comparison of multiple models" with the aim of automatic retraining :blush:

**Comment by KPLauritzen** [comment]
> @KPLauritzen But that is already implemented in scikit-learn? Do you think of it as being a validation step, like "this feature needs to have this positive relationship with prediction" in order for the validation to pass? 😊

I think your comment refers to the Partial dependence plots. 
I think a model evaluator can be many things, and one of them is just plotting a bunch of things that humans can inspect afterwards. So PDPs might not contribute to an automatic validation, but help with understanding what the model learned.
